### PR TITLE
Add MaxRequestBodySize parameter (#478)

### PR DESCRIPTION
**Pull request by cesarbs**
#478

@halter73 @pakrym @natemcmaster

### CODE REVIEW

**Comment by benaadams** [comment]
doit! 💃

**Comment by cesarbs** [comment]
Ready for review.

**Comment by benaadams** [comment]
That's surprisingly clean 😄

File: src/Microsoft.AspNetCore.Server.Kestrel/KestrelServerLimits.cs [hunk 1]
```diff
@@ -23,6 +23,9 @@ public class KestrelServerLimits
         // Matches the default LimitRequestFields in Apache httpd.
         private int _maxRequestHeaderCount = 100;
 
+        // Matches the default maxRequestEntityAllowed in IIS.
+        private long _maxRequestBodySize = 200 * 1024;
```
Comment (natemcmaster) [inline src/Microsoft.AspNetCore.Server.Kestrel/KestrelServerLimits.cs]: Any particular reason for matching IIS instead of nginx or tomcats's default size?

Comment (cesarbs) [inline src/Microsoft.AspNetCore.Server.Kestrel/KestrelServerLimits.cs]: Nope, I just picked one. Needs design I guess. cc @shirhatti

Comment (benaadams) [inline src/Microsoft.AspNetCore.Server.Kestrel/KestrelServerLimits.cs]: Picking one at random; .NET 4-4.6?


**.NET 4 - 4.6** [default](https://msdn.microsoft.com/en-us/library/system.web.configuration.httpruntimesection.maxrequestlength%28v=vs.100%29.aspx) `HttpRuntimeSection.MaxRequestLength` 4MB
**IIS** [default](https://www.iis.net/configreference/system.webserver/asp/limits) `maxRequestEntityAllowed` 200kB
or filtered [default](https://www.iis.net/configreference/system.webserver/security/requestfiltering/requestlimits) 28.6MB _(returns `404.13 Content Length Too Large`)_
**ngnix** [default](http://nginx.org/en/docs/http/ngx_http_core_module.html#client_max_body_size) `client_max_body_size` 1MB
**apache.tomcat** [default](http://tomcat.apache.org/tomcat-9.0-doc/config/http.html) `maxPostSize` 2MB
**apache.httpd** [default](http://httpd.apache.org/docs/2.4/mod/core.html#LimitRequestBody) `LimitRequestBody` unlimited

**Comment by halter73** [comment]
Any tests verifying that the MaxRequestBodySize configuration doesn't affect Upgrade/WebSocket/ForRemainingData requests?

File: src/Microsoft.AspNetCore.Server.Kestrel/KestrelServerLimits.cs (hunk 1 above)
Comment (blowdart) [inline src/Microsoft.AspNetCore.Server.Kestrel/KestrelServerLimits.cs]: As an aside the comment needs to say it's in bytes or kilobytes or megabytes or whatever. Or the variable name needs to state it.

**Comment by cesarbs** [comment]
@halter73 `MaxRequestBodySizeNotEnforcedOnUpgradedConnection`

**Comment by halter73** [comment]
Perfect.

**Comment by halter73** [comment]
:shipit:

File: src/Microsoft.AspNetCore.Server.Kestrel/KestrelServerLimits.cs (hunk 1 above)
Comment (cesarbs) [inline src/Microsoft.AspNetCore.Server.Kestrel/KestrelServerLimits.cs]: @blowdart The comment for the public property states that it's in bytes.

Comment (natemcmaster) [inline src/Microsoft.AspNetCore.Server.Kestrel/KestrelServerLimits.cs]: +1 for 4 MB as default. 200kB seems too small. One common scenario that usually exceeds 200 kB: image uploads.

**Comment by cesarbs** [comment]
Changed default to 4MB.

**Comment by natemcmaster** [comment]
:shipit:

File: src/Microsoft.AspNetCore.Server.Kestrel/KestrelServerLimits.cs [hunk 2]
```diff
@@ -23,6 +23,10 @@ public class KestrelServerLimits
         // Matches the default LimitRequestFields in Apache httpd.
         private int _maxRequestHeaderCount = 100;
 
+        // Matches default HttpRuntimeSection.MaxRequestLength.
+        // https://msdn.microsoft.com/en-us/library/system.web.configuration.httpruntimesection.maxrequestlength%28v=vs.100%29.aspx
+        private long _maxRequestBodySize = 4 * 1024 * 1024;
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Server.Kestrel/KestrelServerLimits.cs]: IIS7 raised this to 28.6MB
https://www.iis.net/configreference/system.webserver/security/requestfiltering/requestlimits

**Comment by Tratcher** [comment]
@cesarbs @muratg this needs an Announcement.

**Comment by halter73** [comment]
@davidfowl Would you remove this change entirely? Or make the default unlimited?

If we want to have endpoint specific configuration, that should probably be middleware. Unfortunately, middleware is almost always opt-in. Even if we update templates, people upgrading existing applications to new packages would still not have request body limits.

**Comment by cesarbs** [comment]
@benaadams Can you tell us why this feature is so important to you (or so it seems 😄)?

**Comment by benaadams** [comment]
@cesarbs I could tell you; but @blowdart wouldn't approve 😉 

I would suggest it is poor practice to have a server that by default allows someone external to it flood its memory infinitely; even if its only via localhost...

**Comment by blowdart** [comment]
So kestrel, as a server, doesn't read the whole request, that's further down the stack, so memory flooding doesn't come into it until it reachs, say, a forms processor. Or so @davidfowl assures me.

**Comment by benaadams** [comment]
Ahh... because `IBufferSizeControl.MaxRequestBufferSize` will throttle it?

**Comment by blowdart** [comment]
I hope that's his thinking :)

**Comment by benaadams** [comment]
May not need this then I suppose...

**Comment by davidfowl** [comment]
Yes please, lets remove it. Sorry @cesarbs 😄

**Comment by cesarbs** [comment]
Bummer.

**Pull request merged**